doozerlib/cli/plashet.py

```diff
@@ -545,15 +550,85 @@ def from_tags(config, brew_tag, embargoed_brew_tag, embargoed_nvr, signing_advis
 
-        if not runtime.assembly:
+        component_builds: Dict[str, Dict] = {}  # candidate rpms for plashet; keys are rpm component names, values are Brew build dicts
+        pinned_nvrs: Dict[str, str] = {}  # rpms pinned to the runtime assembly; keys are rpm component names, values are nvrs
+
+        if not assembly:
             # Assemblies are disabled. We just need the latest tagged builds in the brew tag
             latest_builds = koji_proxy.listTagged(tag, latest=True, inherit=inherit, event=event, type='rpm')
+            component_builds = {build["name"]: build for build in latest_builds}
         else:
             # Assemblies are enabled. We need all tagged builds in the brew tag then find the latest ones for the runtime assembly.
+            if runtime.assembly_basis_event:
+                logger.warning(f'Constraining rpm search to stream assembly due to assembly basis event {runtime.assembly_basis_event}')
+                assembly = 'stream'
             tagged_builds = koji_proxy.listTagged(tag, latest=False, inherit=inherit, event=event, type='rpm')
-            latest_builds = find_latest_builds(tagged_builds, runtime.assembly)
-
-        for build in latest_builds:
+            latest_builds = find_latest_builds(tagged_builds, assembly)
+            component_builds = {build["name"]: build for build in latest_builds}
+
+            el_version_match = re.search(r"rhel-(\d+)", tag)
+            el_version = int(el_version_match[1]) if el_version_match else 0
+            if el_version:  # Only honor pinned rpms if this tag is relevant to a RHEL version
+                # Honor pinned rpm nvrs specified in assembly.members.rpms
+                for component in runtime.releases_config.releases[runtime.assembly].assembly.members.rpms:
+                    nvr = component["is"][f"el{el_version}"]
+                    if not nvr:
+                        continue
+                    if "*" in component.distgit_key:
+                        raise ValueError(f"You cannot pin an NVR in an assembly entry that has a wildcard in its distgit key ({component.distgit_key}).")
+                    rpm_meta = runtime.rpm_map.get(component.distgit_key)
+                    if not rpm_meta:
+                        raise ValueError(f"Distgit key {component.distgit_key} is unknown or excluded.")
+                    nvre_obj = parse_nvr(str(nvr))
+                    if nvre_obj["name"] != rpm_meta.rpm_name:
+                        raise ValueError(f"RPM {nvr} is pinned to assembly {runtime.assembly} for distgit key {component.distgit_key}, but its package name is not {rpm_meta.rpm_name}.")
+                    if isolate_assembly_in_release(nvre_obj["release"]) != runtime.assembly:
+                        raise ValueError(f"RPM {nvr} is pinned to assembly {runtime.assembly}, but that rpm is not a part of the assembly.")
+                    if nvre_obj["name"] in pinned_nvrs and pinned_nvrs[nvre_obj['name']] != nvr:
```

Comment:
I also think this is acceptable in light of assembly inheritance. art1 may define an nvr and art2 can inherit from art1 and override `is` for a component specified by art1.

Reply:
By moving "is" into metadata, I think rpm meta will take care of the assembly inheritance so the rpm name will never repeat. Thoughts?

Reply:
Agreed.